### PR TITLE
GG-28240 Compression alpha

### PR DESCRIPTION
**Pull request by alamar**

### CODE REVIEW

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/GridCacheProcessor.java [hunk 1]
```diff
@@ -2461,6 +2467,21 @@ private CacheGroupContext startCacheGroup(
         boolean persistenceEnabled = recoveryMode || sharedCtx.localNode().isClient() ? desc.persistenceEnabled() :
             dataRegion != null && dataRegion.config().isPersistenceEnabled();
 
+        EntryCompressionStrategy compressionStrategy = null;
+
+        Map<Class<? extends EntryCompressionConfiguration>,
```
Comment (agura) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/GridCacheProcessor.java:2472]: It is really unreadable snippet :) What if we just replace usage below like this:

```java
 if (compressionConfiguration != null) {
            compressionStrategy = CU.entryCompressionStrategies(ctx).get(compressionConfiguration.getClass())
                .apply(compressionConfiguration);

            compressionStrategy.start(ctx, cfg);
 }
```

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/GridCacheUtils.java [hunk 2]
```diff
@@ -2117,6 +2120,29 @@ public static boolean isCacheTemplateName(String cacheName) {
...
+    public static Map<Class<? extends EntryCompressionConfiguration>,
+        IgniteClosure<EntryCompressionConfiguration, EntryCompressionStrategy>>
+        entryCompressionStrategies(GridKernalContext kernalCtx) {
+        Map strategies = new HashMap<>();
+
+        // Adding strategies from plugins.
+        EntryCompressionStrategySupplier[] suppliers = kernalCtx.plugins()
+            .extensions(EntryCompressionStrategySupplier.class);
+
+        if (nonNull(suppliers)) {
```
Comment (agura) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/GridCacheUtils.java:2138]: Please, don't use `nonNull` in such cases. Most code uses `val != null` and there is no any reason for change this style.
Use `nonNull` only as simple lambda-expressions (see corresponding javadoc).

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/tree/CacheDataTree.java [hunk 3]
```diff
@@ -461,38 +462,42 @@ private int compareKeys(KeyCacheObject key, final long link) throws IgniteChecke
                     if (grp.storeCacheIdInDataPage())
                         addr += 4; // Skip cache id.
 
-                    final int len = PageUtils.getInt(addr, 0);
+                    byte type = PageUtils.getByte(addr, 4);
 
-                    int lenCmp = Integer.compare(len, bytes.length);
+                    if (type != CacheObject.TYPE_BINARY_COMPRESSED) {
```
[the remark this answers is not in the document]
Comment (alamar) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/tree/CacheDataTree.java:467]: If type is CacheObject.TYPE_BINARY_COMPRESSED then regular non-optimized comparison takes place, such as with keys larger than page - fetch both entries off PDS, unwrap, compare.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/IgniteCacheOffheapManagerImpl.java [hunk 4]
```diff
@@ -1879,8 +1886,13 @@ private void invoke0(GridCacheContext cctx, CacheSearchRow row, OffheapInvokeClo
                     return false;
 
                 for (GridCacheMvccEntryInfo info : hist) {
+                    CacheObject val = info.value();
+
+                    if (val != null)
+                        val = val.prepareForCache(coCtx, coCtx.compressKeys());
```
[the remark this answers is not in the document]
Comment (alamar) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/IgniteCacheOffheapManagerImpl.java:1892]: Thank you, I will fix.

**Pull request merged**